run/http_runner.go

```diff
@@ -0,0 +1,265 @@
+			QueryParamDecoder[Option],
+			algorithm,
+			GenericEncoder[Solution, Option](encode.JSON()),
+		),
+	}
+
+	runnerConfig := runner.Runner.RunnerConfig()
+	runner.maxParallel = make(chan struct{}, runnerConfig.Runner.HTTP.MaxParallel)
+
+	// default http server
+	runner.httpServer = &http.Server{
+		ReadHeaderTimeout: runnerConfig.Runner.HTTP.ReadHeaderTimeout,
+		Addr:              runnerConfig.Runner.HTTP.Address,
+		ErrorLog:          log.New(os.Stderr, "[Nextmv HTTPRunner] ", log.LstdFlags),
+		Handler:           runner,
+	}
+
+	// default handler to IOProducer
+	runner.httpRequestHandler = SyncHTTPRequestHandler
+
+	for _, option := range options {
+		option(runner)
+	}
+
+	return runner
+}
+
+type httpRunner[Input, Option, Solution any] struct {
+	Runner[HTTPRunnerConfig, Input, Option, Solution]
+	httpServer         *http.Server
+	maxParallel        chan struct{}
+	httpRequestHandler HTTPRequestHandler
+}
+
+func (h *httpRunner[Input, Option, Solution]) setHTTPAddr(addr string) {
+	h.httpServer.Addr = addr
+}
+
+func (h *httpRunner[Input, Option, Solution]) setLogger(l *log.Logger) {
+	h.httpServer.ErrorLog = l
+}
+
+func (h *httpRunner[Input, Option, Solution]) setMaxParallel(maxParallel int) {
+	h.maxParallel = make(chan struct{}, maxParallel)
+}
+
+func (h *httpRunner[Input, Option, Solution]) ActiveRuns() int {
+	return len(h.maxParallel)
+}
+
+func (h *httpRunner[Input, Option, Solution]) setHTTPRequestHandler(
+	f HTTPRequestHandler,
+) {
+	h.httpRequestHandler = f
+}
+
+func (h *httpRunner[Input, Option, Solution]) setHTTPServer(s *http.Server) {
+	h.httpServer = s
+}
+
+func (h *httpRunner[Input, Option, Solution]) setRunnerOption(
+	option RunnerOption[HTTPRunnerConfig, Input, Option, Solution],
+) {
+	option(h.Runner)
+}
+
+func (h *httpRunner[Input, Option, Solution]) Run(
+	context context.Context,
+) error {
+	httpRunnerConfig := h.Runner.RunnerConfig()
+	if httpRunnerConfig.Runner.HTTP.Certificate != "" ||
+		httpRunnerConfig.Runner.HTTP.Key != "" {
+		return h.httpServer.ListenAndServeTLS(
+			httpRunnerConfig.Runner.HTTP.Certificate,
+			httpRunnerConfig.Runner.HTTP.Key,
+		)
+	}
+	return h.httpServer.ListenAndServe()
+}
+
+// ServeHTTP implements the http.Handler interface.
+func (h *httpRunner[Input, Option, Solution]) ServeHTTP(
+	w http.ResponseWriter, req *http.Request,
+) {
+	select {
+	case h.maxParallel <- struct{}{}:
+	default:
+		// No free slot, so we immediately return an error.
+		http.Error(w, "max number of parallel requests exceeded",
+			http.StatusTooManyRequests)
+		return
+	}
+
+	// control mechanism to let the request by run async or not.
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer func() { <-h.maxParallel }()
+		// configure how to turn the request and response into an IOProducer.
+		callbackFunc, producer, err := h.httpRequestHandler(w, req)
+		async := callbackFunc != nil
+		if err != nil {
+			handleError(h.httpServer.ErrorLog, async, err, w)
```

Comment:
There is no `wg.Done()` call here. Doesn't end up in an infinite loop if this error happens?